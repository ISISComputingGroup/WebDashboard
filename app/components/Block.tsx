"use client";
import { IfcBlock } from "@/app/types";
import React, { useState } from "react";

const grafana_stub =
  "https://shadow.nd.rl.ac.uk/grafana/d/wMlwwaHMk/block-history?viewPanel=2&orgId=1&var-block=";

function numberFormatter(value: string | number | undefined) {
  var nValue: number = value == undefined ? NaN : +value
  if (isNaN(nValue)) {
    return value
  } else {
    if (nValue != 0 && (Math.abs(nValue) < 0.001 || Math.abs(nValue) > 10000)) {
      return nValue.toExponential()
    } else {
      return nValue.toPrecision()
    }
  }
}

export default function Block({
  pv,
  instName,
  showHiddenBlocks,
}: {
  pv: IfcBlock;
  instName: string;
  showHiddenBlocks: boolean;
}) {
  const [currentValue, setCurrentValue] = useState<
    string | number | undefined
  >();
  const [showAdvanced, setShowAdvanced] = useState(false);
  if (!pv.visible && !showHiddenBlocks && !instName) {
    return null;
  }
  if (pv.value != currentValue) {
    setCurrentValue(pv.value);
    // Let the user know a change has occurred by lighting up the green dot next to the value
    const pvChangedIndicator = document.getElementById(
      pv.human_readable_name + "_CIRCLE",
    );
    if (!pvChangedIndicator) return;
    if (pvChangedIndicator.classList.contains("text-green-500")) return;
    pvChangedIndicator.classList.remove("text-transparent");
    pvChangedIndicator.classList.add("text-green-500");
    setTimeout(() => {
      pvChangedIndicator.classList.remove("text-green-500");
      pvChangedIndicator.classList.add("text-transparent");
    }, 2000);
  }

  const minimum_date_to_be_shown = 631152000; // This is what PVWS thinks epoch time is for some reason. don't bother showing it as the instrument wasn't running EPICS on 01/01/1990
  return (
    <tr
      key={pv.human_readable_name}
      className="border-b border-blue-gray-200 transition duration-100 hover:bg-gray-100 hover:text-black"
      onClick={() => {
        setShowAdvanced(!showAdvanced);
      }}
    >
      <td className="py-1 px-2 w-1/3 flex-row">
        <a
          className="underline"
          href={
            grafana_stub +
            pv.human_readable_name +
            "&var-inst=" +
            instName.toUpperCase()
          }
          target="_blank"
        >
          {pv.human_readable_name}
        </a>
      </td>

      <td className="py-1 px-2 w-7/12">
        <span id={pv.human_readable_name + "_VALUE_ROW"}>
          <div className="flex justify-between">
            <span
              id={pv.human_readable_name + "_VALUE"}
              className={pv.severity != "NONE" ? "text-red-400" : ""}
            >
              {showAdvanced && "Readback: "}
              {numberFormatter(pv.value)} {pv.units != null && pv.units}
            </span>
            <svg
              id={pv.human_readable_name + "_CIRCLE"}
              className="min-w-2 min-h-2 max-w-2 max-h-2 transition-all text-transparent"
              xmlns="http://www.w3.org/2000/svg"
              fill="currentColor"
              viewBox="0 0 24 24"
            >
              <circle cx="12" cy="12" r="12" />
            </svg>
          </div>

          {showAdvanced && (
            <div>
              <hr />
              {pv.severity != "NONE" ? (
                <a
                  href="https://github.com/ISISComputingGroup/ibex_user_manual/wiki/Blocks#alarms"
                  className="text-red-400"
                  target="_blank"
                >
                  Alarm: {pv.severity}
                </a>
              ) : null}
              <hr />
              {pv.sp_value != null ? (
                <span id={pv.human_readable_name + "_SP"}>
                  {`Setpoint: ${pv.sp_value}`}
                  <hr />
                </span>
              ) : null}
              {pv.updateSeconds != null &&
              pv.updateSeconds > minimum_date_to_be_shown ? (
                <span id={pv.human_readable_name + "_TIMESTAMP"}>
                  {/*Multiply by 1000 here as Date() expects milliseconds*/}
                  {`Last update: ${new Date(pv.updateSeconds * 1000).toLocaleString()}`}
                </span>
              ) : null}
            </div>
          )}
        </span>
      </td>
      <td className="py-1 px-2  flex justify-between items-center">
        <span
          id={pv.human_readable_name + "_VALUE_RC"}
          title={"Run control in-range?"}
        >
          {pv.runcontrol_enabled && (pv.runcontrol_inrange ? "✅" : "❌")}
        </span>

        <span
          className={"cursor-pointer font-bold"}
          title={"Show/Hide advanced statuses"}
        >
          {showAdvanced ? "-" : "+"}
        </span>
      </td>
    </tr>
  );
}
