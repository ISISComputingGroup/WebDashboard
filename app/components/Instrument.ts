import {
  ConfigOutput,
  ConfigOutputBlock,
  DashboardArr,
  IfcBlock,
  IfcGroup,
  IfcPV,
  IfcPVWSMessage,
  IfcPVWSRequest,
  PVWSRequestType,
} from "@/app/types";
import {
  ExponentialOnThresholdFormat,
  findPVByAddress,
} from "@/app/components/PVutils";

const DASHBOARD = "CS:DASHBOARD:TAB:";

export class Instrument {
  prefix: string;
  dashboard_prefix: string;

  groups: Array<IfcGroup> = [];
  runInfoPVs: Array<IfcPV> = [];

  dashboard: DashboardArr = [];

  constructor(prefix: string) {
    this.prefix = prefix;
    this.dashboard_prefix = `${this.prefix}${DASHBOARD}`;

    this.dashboard = [
      //column 0
      [
        [
          { pvaddress: "", value: "Title:" },
          { pvaddress: `${this.prefix}DAE:TITLE` },
        ],
        [
          { pvaddress: "", value: "Users:" },
          { pvaddress: `${this.prefix}DAE:_USERNAME` },
        ],
      ],
      //column 1
      [
        [
          { pvaddress: `${this.dashboard_prefix}1:1:LABEL` },
          { pvaddress: `${this.dashboard_prefix}1:1:VALUE` },
        ],
        [
          { pvaddress: `${this.dashboard_prefix}2:1:LABEL` },
          { pvaddress: `${this.dashboard_prefix}2:1:VALUE` },
        ],
        [
          { pvaddress: `${this.dashboard_prefix}3:1:LABEL` },
          { pvaddress: `${this.dashboard_prefix}3:1:VALUE` },
        ],
      ],
      //column 2
      [
        [
          { pvaddress: `${this.dashboard_prefix}1:2:LABEL` },
          { pvaddress: `${this.dashboard_prefix}1:2:VALUE` },
        ],
        [
          { pvaddress: `${this.dashboard_prefix}2:2:LABEL` },
          { pvaddress: `${this.dashboard_prefix}2:2:VALUE` },
        ],
        [
          { pvaddress: `${this.dashboard_prefix}3:2:LABEL` },
          { pvaddress: `${this.dashboard_prefix}3:2:VALUE` },
        ],
      ],
    ];
    this.runInfoPVs = [
      {
        pvaddress: `${this.prefix}CS:BLOCKSERVER:CURR_CONFIG_NAME`,
        human_readable_name: "Config name",
      },
      {
        pvaddress: `${this.prefix}DAE:RUNSTATE_STR`,
        human_readable_name: "Run state",
      },
      {
        pvaddress: `${this.prefix}DAE:RUNNUMBER`,
        human_readable_name: "Run number",
      },
      { pvaddress: `${this.prefix}DAE:TITLE`, human_readable_name: "Title" },
      {
        pvaddress: `${this.prefix}DAE:_USERNAME`,
        human_readable_name: "Users",
      },
      {
        pvaddress: `${this.prefix}DAE:GOODFRAMES`,
        human_readable_name: "Good frames",
      },
      {
        pvaddress: `${this.prefix}DAE:RAWFRAMES`,
        human_readable_name: "Raw frames",
      },
      {
        pvaddress: `${this.prefix}DAE:COUNTRATE`,
        human_readable_name: "Count Rate",
      },
      {
        pvaddress: `${this.prefix}DAE:_RBNUMBER`,
        human_readable_name: "RB Number",
      },
      {
        pvaddress: `${this.prefix}DAE:BEAMCURRENT`,
        human_readable_name: "Current(uamps)",
      },
      {
        pvaddress: `${this.prefix}DAE:TOTALUAMPS`,
        human_readable_name: "Total(uamps)",
      },
      {
        pvaddress: `${this.prefix}DAE:MONITORCOUNTS`,
        human_readable_name: "Monitor counts",
      },
      {
        pvaddress: `${this.prefix}DAE:MONITORSPECTRUM`,
        human_readable_name: "Monitor Spectrum",
      },
      {
        pvaddress: `${this.prefix}DAE:MONITORFROM`,
        human_readable_name: "Monitor From",
      },
      {
        pvaddress: `${this.prefix}DAE:MONITORTO`,
        human_readable_name: "Monitor To",
      },
      {
        pvaddress: `${this.prefix}CS:SHUTTER`,
        human_readable_name: "Shutter Status",
      },
      {
        pvaddress: `${this.prefix}DAE:NUMSPECTRA`,
        human_readable_name: "Number of Spectra",
      },
      {
        pvaddress: `${this.prefix}DAE:NUMTIMECHANNELS`,
        human_readable_name: "Number of Time Channels",
      },
      {
        pvaddress: `${this.prefix}DAE:SIM_MODE`,
        human_readable_name: "DAE Simulation Mode",
      },
      {
        pvaddress: `${this.prefix}TIME_OF_DAY`,
        human_readable_name: "Instrument Time",
      },
      {
        pvaddress: `${this.prefix}DAE:STARTTIME`,
        human_readable_name: "Start time",
      },
      {
        pvaddress: `${this.prefix}DAE:RUNDURATION`,
        human_readable_name: "Run time(s)",
      },
      { pvaddress: `${this.prefix}DAE:PERIOD`, human_readable_name: "Period" },
      {
        pvaddress: `${this.prefix}DAE:RAWFRAMES_PD`,
        human_readable_name: "Period Raw frames",
      },
      {
        pvaddress: `${this.prefix}DAE:NUMPERIODS`,
        human_readable_name: "Num periods",
      },
      {
        pvaddress: `${this.prefix}DAE:RUNDURATION_PD`,
        human_readable_name: "Period Run Time(s)",
      },
      {
        pvaddress: `${this.prefix}DAE:PERIODSEQ`,
        human_readable_name: "Period Sequence",
      },
      {
        pvaddress: `${this.prefix}DAE:GOODFRAMES_PD`,
        human_readable_name: "Period Good Frames",
      },
      {
        pvaddress: `${this.prefix}DAE:DAEMEMORYUSED`,
        human_readable_name: "DAE Memory Used",
      },
      {
        pvaddress: `${this.prefix}DAE:DAETIMINGSOURCE`,
        human_readable_name: "Timing Source",
      },
      {
        pvaddress: `${this.prefix}DAE:EVENTS`,
        human_readable_name: "Total Counts",
      },
    ];
  }
}

export function findPVInDashboard(
  dashboard: DashboardArr,
  pvAddress: string,
): undefined | IfcPV {
  return findPVByAddress(dashboard.flat(3), pvAddress);
}

export const RC_ENABLE = ":RC:ENABLE";
export const RC_INRANGE = ":RC:INRANGE";
export const SP_RBV = ":SP:RBV";
export const CSSB = "CS:SB:";

export function toPrecision(
  block: IfcPV,
  pvVal: number | string,
): string | number {
  return block.precision
    ? ExponentialOnThresholdFormat(pvVal, block.precision)
    : pvVal;
}

export function storePrecision(updatedPV: IfcPVWSMessage, block: IfcBlock) {
  const prec = updatedPV.precision;
  if (prec != null && prec > 0 && !block.precision) {
    // this is likely the first update, and contains precision information which is not repeated on a normal value update - store this in the block for later truncation (see below)
    block.precision = prec;
  }
}

export function yesToBoolean(pvVal: string | number) {
  return pvVal == "YES";
}

export function subscribeToBlockPVs(
  sendJsonMessage: (a: IfcPVWSRequest) => void,
  block_address: string,
) {
  /**
   * Subscribes to a block and its associated run control PVs
   */
  sendJsonMessage({
    type: PVWSRequestType.subscribe,
    pvs: [
      block_address,
      block_address + RC_ENABLE,
      block_address + RC_INRANGE,
      block_address + SP_RBV,
    ],
  });
}

/**
 * Parse the blockserver's current configuration output
 * and create an array of groups which contain blocks.
 */
export function getGroupsWithBlocksFromConfigOutput(
  configOutput: ConfigOutput,
  sendJsonMessage: (a: IfcPVWSRequest) => void,
  prefix: string,
): Array<IfcGroup> {
  const groups = configOutput.groups;
  let newGroups: Array<IfcGroup> = [];
  for (const group of groups) {
    const groupName = group.name;
    let blocks: Array<IfcBlock> = [];
    for (const block of group.blocks) {
      const newBlock = configOutput.blocks.find(
        (b: ConfigOutputBlock) => b.name === block,
      );
      if (newBlock) {
        blocks.push({
          pvaddress: newBlock.pv,
          human_readable_name: newBlock.name,
          low_rc: newBlock.lowlimit,
          high_rc: newBlock.highlimit,
          visible: newBlock.visible,
        });
        const fullyQualifiedBlockPVAddress = prefix + CSSB + newBlock.name;
        subscribeToBlockPVs(sendJsonMessage, fullyQualifiedBlockPVAddress);
      }
    }
    newGroups.push({
      name: groupName,
      blocks: blocks,
    });
  }
  return newGroups;
}

export function findPVInGroups(
  groups: Array<IfcGroup>,
  prefix: string,
  updatedPVName: string,
): undefined | IfcBlock {
  return groups
    .flatMap((group: IfcGroup) => group.blocks)
    .find(
      (block: IfcBlock) =>
        updatedPVName == prefix + CSSB + block.human_readable_name,
    );
}
