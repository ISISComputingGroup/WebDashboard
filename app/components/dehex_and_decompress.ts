import pako from "pako";

function unhexlify(str: string): string {
  let result = "";
  for (let i = 0, l = str.length; i < l; i += 2) {
    result += String.fromCharCode(parseInt(str.substr(i, 2), 16));
  }
  return result;
}

/**
 * dehex_and_decompress
 * - synonymous to dehex_and_decompress in inst_servers
 * @param {*} input raw data
 * @returns dehexed and decompressed data (you can choose to JSON parse it or not afterwards)
 */
export function dehex_and_decompress(
  input: string,
): string | Uint8Array | null {
  // DEHEX
  const unhexed = unhexlify(input);
  const charData = unhexed.split("").map(function (x) {
    return x.charCodeAt(0);
  });
  // convert to binary
  const binData = new Uint8Array(charData);
  return pako.inflate(binData, { to: "string" });
}
