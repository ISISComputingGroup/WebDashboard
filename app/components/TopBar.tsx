import {
  getForegroundColour,
  getStatusColour,
  UNREACHABLE,
} from "./getRunstateColours";

import { tBlockMapping } from "@/app/types";
import { DASHBOARD } from "@/app/components/Instrument";
import { memo } from "react";

export function getRunstate(prefix: string, runInfoPVs: tBlockMapping): string {
  const runStatePV = runInfoPVs.get(`${prefix}DAE:RUNSTATE_STR`);
  if (runStatePV && runStatePV.value && typeof runStatePV.value === "string") {
    return runStatePV.value;
  }
  return UNREACHABLE;
}

const TopBar = function TopBar({
  dashboard,
  instName,
  runInfoPVs,
  prefix,
}: {
  dashboard: tBlockMapping;
  instName: string;
  runInfoPVs: tBlockMapping;
  prefix: string;
}) {
  if (!dashboard || !runInfoPVs || !instName) {
    return (
      <h1 className="text-lg w-full text-white bg-gray-400 border-gray-500 border-2 p-3 font-semibold px-7 animate-pulse">
        Loading...
      </h1>
    );
  }

  return (
    <div
      id="top_bar"
      className="w-full bg-white  shadow-lg text-black rounded-xl text-md"
    >
      <div
        id="instNameDiv"
        className="w-full flex justify-center items-center flex-col"
      >
        <div
          id={"runstateDiv"}
          className={`p-2 rounded-t-lg w-full flex-col justify-center items-center text-center ${getStatusColour(getRunstate(prefix, runInfoPVs))} ${getForegroundColour(getRunstate(prefix, runInfoPVs))}`}
        >
          <h2 className={"text-xl"} id={"instNameLabel"}>
            {instName.toUpperCase()} is{" "}
            <span id={"runStateSpan"}>{getRunstate(prefix, runInfoPVs)}</span>
          </h2>
          <div
            id={"banner"}
            className={"flex flex-col md:flex-row text-center w-full"}
          >
            {["LEFT", "MIDDLE", "RIGHT"].map((place) => (
              <div key={place} className={"flex-row md:w-1/3"}>
                {
                  dashboard.get(`${prefix}CS:DASHBOARD:BANNER:${place}:LABEL`)!
                    .value
                }{" "}
                {
                  dashboard.get(`${prefix}CS:DASHBOARD:BANNER:${place}:VALUE`)!
                    .value
                }
              </div>
            ))}
          </div>
        </div>
        <h3 className="text-black text-wrap max-w-full break-all py-2 font-semibold">
          Config:{" "}
          <span id={"configNameSpan"}>
            {runInfoPVs.get(`${prefix}CS:BLOCKSERVER:CURR_CONFIG_NAME`)
              ? runInfoPVs.get(`${prefix}CS:BLOCKSERVER:CURR_CONFIG_NAME`)!
                  .value
              : "UNKNOWN"}
          </span>
        </h3>

        <div className="bg-gray-50 border-2 border-gray-200 flex flex-col max-w-full w-full">
          <table
            id={"dashboardTable"}
            className="text-sm max-w-full table-fixed flex divide-x divide-gray-200 text-wrap w-full "
          >
            <tbody className="w-1/3">
              <tr className="not-last:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex">
                <td className="py-1 px-4 flex font-bold break-words">Title</td>
                <td className="py-1 px-4 flex justify-between items-center break-all">
                  <span className="font-light">
                    {dashboard.get(`${prefix}DAE:WDTITLE`)?.value
                      ? dashboard.get(`${prefix}DAE:WDTITLE`)!.value
                      : "Unknown"}
                  </span>
                </td>
              </tr>

              <tr className="not-last:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex">
                <td className="py-1 px-4 flex font-bold break-words">Users</td>
                <td className="py-1 px-4 flex justify-between items-center break-all">
                  <span className="font-light">
                    {dashboard.get(`${prefix}DAE:WDUSERS`)?.value
                      ? dashboard.get(`${prefix}DAE:WDUSERS`)!.value
                      : "Unknown"}
                  </span>
                </td>
              </tr>
            </tbody>

            {[1, 2].map((col: number) => (
              <tbody key={col} id={col.toString()} className="w-1/3">
                {[1, 2, 3].map((row: number) => (
                  <tr
                    key={row}
                    className="not-last:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
                  >
                    <td className="py-1 px-4 flex font-bold break-words">
                      {
                        dashboard.get(
                          `${prefix}${DASHBOARD}${row}:${col}:LABEL`,
                        )!.value
                      }
                    </td>
                    <td className="py-1 px-4 flex justify-between items-center break-all">
                      <span className="font-light">
                        {dashboard.get(
                          `${prefix}${DASHBOARD}${row}:${col}:VALUE`,
                        )?.value
                          ? dashboard.get(
                              `${prefix}${DASHBOARD}${row}:${col}:VALUE`,
                            )!.value
                          : "Unknown"}
                      </span>
                    </td>
                  </tr>
                ))}
              </tbody>
            ))}
          </table>
        </div>

        <label id={"runInfoLabel"} className="p-2 w-full">
          <input className="peer/showLabel absolute scale-0" type="checkbox" />
          <span className="block max-h-14 overflow-hidden rounded-lg bg-gray-600 hover:bg-gray-800 text-white px-4 py-0 mb-2 transition-all duration-300 peer-checked/showLabel:max-h-fit cursor-pointer items-center">
            <h3 className="flex h-14 cursor-pointer font-bold text-center items-center justify-center">
              Show/hide all run information
            </h3>
            <div className="grid md:grid-cols-5 gap-2">
              {Array.from(runInfoPVs.values()).map((runInfoPV) => (
                <div
                  className="mb-2 shadow-xs rounded-md"
                  key={runInfoPV.human_readable_name}
                >
                  <p className="font-bold">{runInfoPV.human_readable_name}:</p>{" "}
                  <p className="break-all">{runInfoPV.value}</p>
                </div>
              ))}
            </div>
          </span>
        </label>
      </div>
    </div>
  );
};

export default memo(TopBar);
export const exportedForTesting = TopBar;
