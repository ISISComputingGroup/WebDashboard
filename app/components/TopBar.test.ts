import TopBar, { getRunstate, runStateStr } from "@/app/components/TopBar";
import { tBlockMapping } from "@/app/types";
import { render } from "@testing-library/react";
import { Instrument } from "@/app/components/Instrument";

test("GetRunstate returns the runstate when it exists and is of string type", () => {
  const prefix = "TESTING:";

  const expected = "SETUP";
  let blocks: tBlockMapping = new Map();
  blocks.set(prefix + "DAE:RUNSTATE_STR", {
    pvaddress: prefix + "DAE:RUNSTATE_STR",
    value: expected,
    human_readable_name: runStateStr,
  });
  expect(getRunstate(prefix, blocks)).toBe(expected);
});

test("GetRunstate returns unknown when no runstate PV in array", () => {
  expect(getRunstate("", new Map())).toBe("UNREACHABLE");
});

it("renders topbar unchanged", () => {
  const prefix = "TESTING:";

  let instrument = new Instrument(prefix);
  const instName = "Instrument";
  const { container } = render(
    TopBar({
      dashboard: instrument.dashboard,
      instName: instName,
      runInfoPVs: instrument.runInfoPVs,
      prefix: prefix,
    }),
  );
  expect(container).toMatchSnapshot();
});

it("draws instName expectedly", () => {
  const prefix = "TESTING:";

  let instrument = new Instrument(prefix);
  const instName = "Instrument";
  const { container } = render(
    TopBar({
      dashboard: instrument.dashboard,
      instName: instName,
      runInfoPVs: instrument.runInfoPVs,
      prefix: prefix,
    }),
  );
  expect(container.querySelector("#instNameLabel")!.innerHTML).toContain(
    instName.toUpperCase(),
  );
});

it("draws configName expectedly", () => {
  const prefix = "TESTING:";
  let instrument = new Instrument(prefix);
  let configNamePV = instrument.runInfoPVs.get(
    prefix + "CS:BLOCKSERVER:CURR_CONFIG_NAME",
  );
  const expectedConfigName = "Aconfig";
  configNamePV!.value = expectedConfigName;
  const { container } = render(
    TopBar({
      dashboard: instrument.dashboard,
      instName: "Instrument",
      runInfoPVs: instrument.runInfoPVs,
      prefix: prefix,
    }),
  );
  expect(container.querySelector("#configNameSpan")!.innerHTML).toBe(
    expectedConfigName,
  );
});
