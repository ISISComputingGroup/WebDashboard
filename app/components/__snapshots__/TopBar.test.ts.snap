// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders topbar unchanged 1`] = `
<div>
  <div
    class="w-full bg-white  shadow-lg text-black rounded-xl text-md"
    id="top_bar"
  >
    <div
      class="w-full flex justify-center items-center flex-col"
      id="instNameDiv"
    >
      <h2
        class="text-center p-4 text-xl rounded-t-lg w-full 
          bg-[#F08080]
           
           text-black
          
          "
        id="instNameLabel"
      >
        INSTRUMENT
         is
         
        <span
          id="runStateSpan"
        >
          UNKNOWN
        </span>
      </h2>
      <h3
        class="text-black text-wrap max-w-full break-all py-2 font-semibold"
      >
        Config:
         
        <span
          id="configNameSpan"
        />
      </h3>
      <div
        class="bg-gray-50 border-2 border-gray-200 flex flex-col max-w-full w-full"
      >
        <table
          class="text-sm max-w-full table-fixed flex divide-x divide-gray-200 text-wrap "
          id="dashboardTable"
        >
          <tbody
            class="w-1/3"
            id="0"
          >
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              >
                Title:
              </td>
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              >
                Users:
              </td>
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
          </tbody>
          <tbody
            class="w-1/3"
            id="1"
          >
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              />
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              />
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              />
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
          </tbody>
          <tbody
            class="w-1/3"
            id="2"
          >
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              />
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              />
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
            <tr
              class="[&:not(:last-child)]:border-b border-gray-300 text-black transition duration-100 hover:bg-gray-700 hover:text-white md:flex"
            >
              <td
                class="py-1 px-4 flex font-bold break-words"
              />
              <td
                class="py-1 px-4 flex justify-between items-center break-all"
              >
                <span
                  class="font-light"
                >
                  Hidden/unknown
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <label
        class="p-2 w-full"
        id="runInfoLabel"
      >
        <input
          class="peer/showLabel absolute scale-0"
          type="checkbox"
        />
        <span
          class="block max-h-14 overflow-hidden rounded-lg bg-gray-600 hover:bg-gray-800 text-white px-4 py-0 mb-2 transition-all duration-300 peer-checked/showLabel:max-h-fit cursor-pointer items-center"
        >
          <h3
            class="flex h-14 cursor-pointer font-bold text-center items-center justify-center"
          >
            Show/hide all run information
          </h3>
          <div
            class="grid md:grid-cols-5 gap-2"
          >
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Config name
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Run state
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Run number
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Title
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Users
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Good frames
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Raw frames
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Count Rate
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                RB Number
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Current(uamps)
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Total(uamps)
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Monitor counts
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Monitor Spectrum
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Monitor From
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Monitor To
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Shutter Status
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Number of Spectra
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Number of Time Channels
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                DAE Simulation Mode
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Instrument Time
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Start time
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Run time(s)
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Period
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Period Raw frames
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Num periods
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Period Run Time(s)
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Period Sequence
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Period Good Frames
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                DAE Memory Used
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Timing Source
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
            <div
              class="mb-2 shadow-sm rounded-md"
            >
              <p
                class="font-bold"
              >
                Total Counts
                :
              </p>
               
              <p
                class="break-all"
              />
            </div>
          </div>
        </span>
      </label>
    </div>
  </div>
</div>
`;
